core/meta/src/TClass.cxx

```diff
@@ -4359,29 +4359,33 @@ Int_t TClass::GetNmethods()
 
 TVirtualStreamerInfo* TClass::GetStreamerInfo(Int_t version /* = 0 */) const
 {
-   TVirtualStreamerInfo *guess = fLastReadInfo;
-   if (guess && guess->GetClassVersion() == version) {
-      // If the StreamerInfo is assigned to the fLastReadInfo, we are
-      // guaranteed it was built and compiled.
-      return guess;
-   }
+   TVirtualStreamerInfo *sinfo = fLastReadInfo;
 
-   R__LOCKGUARD(gInterpreterMutex);
+   // Version 0 is special, it means the currently loaded version.
+   // We need to set it at the beginning to be able to guess it correctly.
 
-   // Handle special version, 0 means currently loaded version.
-   // Warning:  This may be -1 for an emulated class.
-   // If version == -2, the user is requested the emulated streamerInfo
-   // for an abstract base class even though we have a dictionary for it.
-   if (version == 0) {
+   if (version == 0)
       version = fClassVersion;
-   }
-   Int_t ninfos = fStreamerInfo->GetSize();
-   if ((version < -1) || (version >= ninfos)) {
+
+   // If the StreamerInfo is assigned to the fLastReadInfo, we are
+   // guaranteed it was built and compiled.
+   if (sinfo && sinfo->GetClassVersion() == version)
+      return sinfo;
+
+   R__LOCKGUARD(gInterpreterMutex);
+
+   // Warning: version may be -1 for an emulated class, or -2 if the
+   //          user requested the emulated streamerInfo for an abstract
+   //          base class, even though we have a dictionary for it.
+
+   if ((version < -1) || (version >= fStreamerInfo->GetSize())) {
       Error("GetStreamerInfo", "class: %s, attempting to access a wrong version: %d", GetName(), version);
       // FIXME: Shouldn't we go to -1 here, or better just abort?
-      version = 0;
+      version = fClassVersion;
```

Comment:
That change is a functional one, right? I don't see it motivated by the commit message ("Avoid lookup for StreamerInfo when it is cached"); could you maybe even move this into a separate commit explaining why this change should happen? (I am not questioning that it's a good thing, I bet you discussed with Philippe, but these changes are not obvious to pinpoint when looking at commit logs.)

Reply:
I need @pcanal  to take a look, actually, that's why I included this commit again. Let's see what he says. If `version = 0` is used, then it will check for the `StreamerInfo` every time. The version itself is cached in `fClassVersion`, so you can only avoid the lookup when it's already there if you check against the same number. `version = 0` means current, which is whatever is stored in `fClassVersion`.

Reply:
> version = 0 means current, 

Technically, no anymore at this point in the code.  It now means literally the one in slot zero .. but there is never any StreamerInfo in that slot, so it seems it leads to the execution of line 4396 and hence the leak of a TStreamerInfo (in cases where the fClassVersion's StreamerInfo was already build.

This code snippet is handling the error case (an non-existing StreamerInfo was asked for) and the new value of 'version' was essentially arbitrary and Guilherme's value is an improvement.

Reply:
Anyway, I agree with Axel that this should be a separate commit ;)  Thanks.